File: tools/math_dsl/expanding_symbol.py
"""W244 wave 18 — closed-form analytical model for `expanding_symbol` FS.

Industry pattern (Novomatic Book of Ra, Play'n GO Book of Dead, all
"Book of …" clones, Microgaming Book of Atem, Spinomenal Book of Demi
Gods, Pragmatic Book of Tut):

  Trigger
  -------
    Free Spins trigger by ≥ N scatter (typical N=3 on 5×3 grid).
    Award: `fs_initial_spins` (typical 10).

  Expansion symbol selection
  --------------------------
    At FS start, one EXPANDING SYMBOL is randomly drawn from the
    paytable's HP set with weighted probability (per published
    `expansion_symbol_table`). The chosen symbol stays fixed for the
    entire FS episode.

  Per-FS-spin payout
  ------------------
    Each spin evaluates lines + expansions:
      * Standard line pays (normal grid)
      * Plus: every reel with ≥ 1 instance of the expanding symbol
        EXPANDS to fill all rows on that reel, then awards as
        pay-anywhere across rows.

  Closed-form RTP contribution (FS portion alone)
  -----------------------------------------------
    Per-FS-spin expansion expected value:
      p_per_reel = 1 - (1 - p_cell) ^ rows
      Σ_reels_active = sum over Bernoulli of reel-expansion outcomes
                     ≈ reels × p_per_reel

      E[expansion_pay_per_spin] = sum_symbols(
          weight_symbol/total_w ×
          (reels × p_per_reel × pay_anywhere_award[symbol] +
           normal_line_pay[symbol])
      )

    For just the EXPANSION component (excluding normal line pays):
      E[expansion_only] = sum_syms(weight/total × Σ_reels_expand_pay)

    Per-trigger total:
      E[trigger_total] = fs_initial_spins × E[expansion_pay_per_spin]

    Per-base-spin RTP contribution:
      RTP[fs_expansion] = P(FS_trigger_per_base_spin) × E[trigger_total]

Pure-stdlib. Used by:
  * `tools.math_dsl.compile` for IR emission
  * `tools/build_expanding_symbol_kernel.py` for acceptance artefact
  * `tools/tests/test_w244_expanding_symbol_kernel.py` for closed-form pin
"""
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class ExpandingSymbolParams:
    """Closed-form model inputs."""
    # FS trigger probability per base spin
    fs_trigger_p: float
    fs_initial_spins: int                   # typical 10
    # Grid topology
    reels: int                              # typical 5
    rows: int                               # typical 3
    # Per-cell landing probability for the expanding symbol.
    # Single value because the FS reels are typically modified to inflate
    # the expanding-symbol weight. Public-spec proxies use 0.08-0.15 range.
    p_per_cell_in_fs: float
    # Pay-anywhere award for full-reel expansion. With `rows` rows on a
    # reel filled with the symbol, this is the published pay schedule:
    # `pay_anywhere_x_bet[reels_expanded]` (table indexed by 1-5 reels).
    # Typical Book-style: {1: 0, 2: 0, 3: 1, 4: 5, 5: 100} or similar.
    pay_table: dict[int, float]
    # Optional: name of the chosen expansion symbol (for audit).
    symbol_name: str = "?"

    def __post_init__(self):
        if not (0.0 <= self.fs_trigger_p <= 1.0):
            raise ValueError(f"fs_trigger_p {self.fs_trigger_p} outside [0,1]")
        if self.fs_initial_spins < 1:
            raise ValueError("fs_initial_spins must be ≥ 1")
        if self.reels < 1:
            raise ValueError("reels must be ≥ 1")
        if self.rows < 1:
            raise ValueError("rows must be ≥ 1")
        if not (0.0 <= self.p_per_cell_in_fs <= 1.0):
            raise ValueError(
                f"p_per_cell_in_fs {self.p_per_cell_in_fs} outside [0,1]"
            )
        if not self.pay_table:
            raise ValueError("pay_table must be non-empty")
        for k, v in self.pay_table.items():
            if k < 0:
                raise ValueError(f"pay_table key {k} must be ≥ 0")
            if v < 0:
                raise ValueError(f"pay_table value {v} must be ≥ 0")


def reel_expansion_probability(p_per_cell: float, rows: int) -> float:
    """P(at least one expanding symbol lands on a reel of `rows` cells).

    Each cell is Bernoulli(p_per_cell), so:
        P(≥1 on reel) = 1 - (1 - p_per_cell) ^ rows
    """
    if p_per_cell >= 1.0:
        return 1.0
    return 1.0 - (1.0 - p_per_cell) ** rows


def expected_reels_expanded(
    p_per_cell: float, reels: int, rows: int
) -> float:
    """E[number of reels with ≥1 expanding symbol].

    By linearity of expectation: reels × P(at least one on a reel).
    """
    return reels * reel_expansion_probability(p_per_cell, rows)


def expected_pay_per_fs_spin(params: ExpandingSymbolParams) -> float:
    """E[expansion pay × bet | one FS spin].

    Computes the full distribution over (k_reels_expanded ∈ [0, reels])
    via Binomial(reels, p_per_reel) and weights by `pay_table[k]`.
    """
    p_per_reel = reel_expansion_probability(
        params.p_per_cell_in_fs, params.rows
    )
    expected = 0.0
    # Binomial PMF over k_reels = 0..reels
    n = params.reels
    q = 1.0 - p_per_reel
    if q == 0:
        # p == 1 → always all reels expand
        return params.pay_table.get(n, 0.0)
    # Iterative PMF construction
    pmf = q ** n  # k = 0
    expected += pmf * params.pay_table.get(0, 0.0)
    for k in range(1, n + 1):
        pmf *= (n - k + 1) / k * (p_per_reel / q)
        expected += pmf * params.pay_table.get(k, 0.0)
    return expected


def expected_pay_per_trigger(params: ExpandingSymbolParams) -> float:
    """E[expansion pay × bet | one FS trigger]."""
    return params.fs_initial_spins * expected_pay_per_fs_spin(params)


def expanding_symbol_rtp(params: ExpandingSymbolParams) -> dict:
    """Per-base-spin RTP contribution + audit breakdown."""
    p_per_reel = reel_expansion_probability(
        params.p_per_cell_in_fs, params.rows
    )
    e_reels = expected_reels_expanded(
        params.p_per_cell_in_fs, params.reels, params.rows
    )
    e_pay_per_spin = expected_pay_per_fs_spin(params)
    e_pay_per_trigger = expected_pay_per_trigger(params)
    rtp = params.fs_trigger_p * e_pay_per_trigger
    return {
        "rtp_contribution": rtp,
        "fs_trigger_p": params.fs_trigger_p,
        "fs_initial_spins": params.fs_initial_spins,
        "reels": params.reels,
        "rows": params.rows,
        "p_per_cell_in_fs": params.p_per_cell_in_fs,
        "p_per_reel": p_per_reel,
        "expected_reels_expanded_per_spin": e_reels,
        "expected_pay_per_fs_spin": e_pay_per_spin,
        "expected_pay_per_trigger": e_pay_per_trigger,
        "symbol_name": params.symbol_name,
        "pay_table": dict(params.pay_table),
    }

File: rust-sim/src/kernels/expanding_symbol.rs
//! W244.18 — expanding_symbol Rust port.
//!
//! Mirror of `tools/math_dsl/expanding_symbol.py`. Closed-form:
//!   P(≥1 on reel)     = 1 − (1 − p_per_cell) ^ rows
//!   E[reels_expanded] = reels × P(≥1 on reel)
//!   E[pay/FS spin]    = Σ_k Binomial(reels, p_per_reel)(k) × pay_table[k]
//!   E[pay/trigger]    = fs_initial_spins × E[pay/FS spin]
//!   RTP               = fs_trigger_p × E[pay/trigger]
//!
//! Used by Book of Dead, Book of Ra, all "book"-style expanding-symbol
//! Free Spins games.

use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExpandingSymbolParams {
    pub fs_trigger_p: f64,
    pub fs_initial_spins: u32,
    pub reels: u32,
    pub rows: u32,
    pub p_per_cell_in_fs: f64,
    pub pay_table: BTreeMap<u32, f64>,
    #[serde(default = "default_symbol_name")]
    pub symbol_name: String,
}

fn default_symbol_name() -> String {
    "?".to_string()
}

impl ExpandingSymbolParams {
    pub fn validate(&self) -> Result<(), String> {
        if !(0.0..=1.0).contains(&self.fs_trigger_p) {
            return Err(format!(
                "fs_trigger_p {} outside [0,1]",
                self.fs_trigger_p
            ));
        }
        if self.fs_initial_spins < 1 {
            return Err("fs_initial_spins must be ≥ 1".into());
        }
        if self.reels < 1 {
            return Err("reels must be ≥ 1".into());
        }
        if self.rows < 1 {
            return Err("rows must be ≥ 1".into());
        }
        if !(0.0..=1.0).contains(&self.p_per_cell_in_fs) {
            return Err(format!(
                "p_per_cell_in_fs {} outside [0,1]",
                self.p_per_cell_in_fs
            ));
        }
        if self.pay_table.is_empty() {
            return Err("pay_table must be non-empty".into());
        }
        for (&k, &v) in &self.pay_table {
            if v < 0.0 {
                return Err(format!("pay_table[{}]={} must be ≥ 0", k, v));
            }
        }
        Ok(())
    }
}

pub fn reel_expansion_probability(p_per_cell: f64, rows: u32) -> f64 {
    if p_per_cell >= 1.0 {
        return 1.0;
    }
    1.0 - (1.0 - p_per_cell).powi(rows as i32)
}

pub fn expected_reels_expanded(p_per_cell: f64, reels: u32, rows: u32) -> f64 {
    (reels as f64) * reel_expansion_probability(p_per_cell, rows)
}

pub fn expected_pay_per_fs_spin(params: &ExpandingSymbolParams) -> f64 {
    let p_per_reel = reel_expansion_probability(params.p_per_cell_in_fs, params.rows);
    let n = params.reels;
    let q = 1.0 - p_per_reel;
    if q == 0.0 {
        return *params.pay_table.get(&n).unwrap_or(&0.0);
    }
    let mut expected = 0.0_f64;
    let mut pmf = q.powi(n as i32);
    expected += pmf * params.pay_table.get(&0).unwrap_or(&0.0);
    for k in 1..=n {
        pmf *= ((n - k + 1) as f64) / (k as f64) * (p_per_reel / q);
        expected += pmf * params.pay_table.get(&k).unwrap_or(&0.0);
    }
    expected
}

pub fn expected_pay_per_trigger(params: &ExpandingSymbolParams) -> f64 {
    (params.fs_initial_spins as f64) * expected_pay_per_fs_spin(params)
}

#[derive(Debug, Serialize)]
pub struct ExpandingSymbolResult {
    pub rtp_contribution: f64,
    pub fs_trigger_p: f64,
    pub fs_initial_spins: u32,
    pub reels: u32,
    pub rows: u32,
    pub p_per_cell_in_fs: f64,
    pub p_per_reel: f64,
    pub expected_reels_expanded_per_spin: f64,
    pub expected_pay_per_fs_spin: f64,
    pub expected_pay_per_trigger: f64,
    pub symbol_name: String,
    pub pay_table: BTreeMap<u32, f64>,
}

pub fn expanding_symbol_rtp(params: &ExpandingSymbolParams) -> ExpandingSymbolResult {
    let p_per_reel = reel_expansion_probability(params.p_per_cell_in_fs, params.rows);
    let e_reels = expected_reels_expanded(params.p_per_cell_in_fs, params.reels, params.rows);
    let e_pay_per_spin = expected_pay_per_fs_spin(params);
    let e_pay_per_trigger = expected_pay_per_trigger(params);
    let rtp = params.fs_trigger_p * e_pay_per_trigger;
    ExpandingSymbolResult {
        rtp_contribution: rtp,
        fs_trigger_p: params.fs_trigger_p,
        fs_initial_spins: params.fs_initial_spins,
        reels: params.reels,
        rows: params.rows,
        p_per_cell_in_fs: params.p_per_cell_in_fs,
        p_per_reel,
        expected_reels_expanded_per_spin: e_reels,
        expected_pay_per_fs_spin: e_pay_per_spin,
        expected_pay_per_trigger: e_pay_per_trigger,
        symbol_name: params.symbol_name.clone(),
        pay_table: params.pay_table.clone(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn book_of_dead_like() -> ExpandingSymbolParams {
        let mut pt = BTreeMap::new();
        pt.insert(0, 0.0);
        pt.insert(1, 0.0);
        pt.insert(2, 0.0);
        pt.insert(3, 1.0);
        pt.insert(4, 5.0);
        pt.insert(5, 100.0);
        ExpandingSymbolParams {
            fs_trigger_p: 0.005,
            fs_initial_spins: 10,
            reels: 5,
            rows: 3,
            p_per_cell_in_fs: 0.10,
            pay_table: pt,
            symbol_name: "S_HP1".into(),
        }
    }

    #[test]
    fn p_per_reel_matches_bernoulli_formula() {
        // P(≥1 in 3 cells, each p=0.1) = 1 − 0.9^3 = 0.271
        let p = reel_expansion_probability(0.10, 3);
        assert!((p - 0.271).abs() < 1e-12);
    }

    #[test]
    fn p_per_cell_one_returns_one_reel_prob() {
        assert_eq!(reel_expansion_probability(1.0, 3), 1.0);
    }

    #[test]
    fn e_reels_expanded_linearity_check() {
        let e = expected_reels_expanded(0.10, 5, 3);
        assert!((e - 5.0 * 0.271).abs() < 1e-12);
    }

    #[test]
    fn rtp_contribution_positive_on_book_pattern() {
        let p = book_of_dead_like();
        let r = expanding_symbol_rtp(&p);
        assert!(r.rtp_contribution > 0.0);
        assert!(r.expected_pay_per_trigger > 0.0);
    }

    #[test]
    fn validate_rejects_zero_spins() {
        let mut pt = BTreeMap::new();
        pt.insert(3, 1.0);
        let p = ExpandingSymbolParams {
            fs_trigger_p: 0.01, fs_initial_spins: 0,
            reels: 5, rows: 3, p_per_cell_in_fs: 0.1,
            pay_table: pt, symbol_name: "x".into(),
        };
        assert!(p.validate().is_err());
    }
}
